tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -1542,6 +1542,46 @@ template <typename T> T waveMultiPrefixProduct(T A, UINT) {
   return A * A;
 }
 
+template <typename T> struct Op<OpType::WaveMatch, T, 1> : StrictValidation {};
+
+template <typename T> struct ExpectedBuilder<OpType::WaveMatch, T> {
+  static std::vector<UINT> buildExpected(Op<OpType::WaveMatch, T, 1> &,
+                                         const InputSets<T> &,
+                                         UINT WaveSize) {
+    // For this test, the shader arranges it so that lane 0 is different from
+    // all the other lanes. Besides that all other lines write their result of
+    // WaveMatch as well.
+
+    std::vector<UINT> Expected;
+    Expected.assign(WaveSize * 4, 0);
+
+    UINT LowWaves = std::min(64U, WaveSize);
+    UINT HighWaves = WaveSize - LowWaves;
+
+    uint64_t LowWaveShift = (LowWaves < 64) ? (1ULL << LowWaves) : 0;
+    uint64_t HighWaveShift = (HighWaves < 64) ? (1ULL << HighWaves) : 0;
+
+    uint64_t result[2] = {(LowWaveShift - 1 & ~1ULL),
+                          (HighWaveShift - 1 & ~0ULL)};
```

Comment:
I've been trying to nudge you in (what I think is) the right direction here.  It really bothers me that the "build the mask" and the "what we expect the result to be" are mangled into the same expression.  This is my recommended approach that separates the concerns:

```suggestion
    uint64_t LowWaveMask = ((LowWaves < 64) ? (1ULL << LowWaves) : 0) - 1;
    uint64_t HighWaveMask = ((HighWaves < 64) ? (1ULL << HighWaves) : 0) - 1;

    uint64_t LowExpected = ~1ULL & LowWaveMask;
    uint64_t HighExpected = ~0ULL & HighWaveMask;
```

IMO, separating it like this makes it clearer what the different calculations are.

Although this is take-it-or-leave-it, I do really think you should factor it this way!